cookbook/security/guard-api-key.rst

```diff
@@ -0,0 +1,489 @@
+.. index::
+    single: Security; Custom Request Authenticator
+
+How to Authenticate Users with an API Key using Guard
+=====================================================
+
+Nowadays, it's quite usual to authenticate the user via an API key (when developing
+a web service for instance). The API key is provided for every request and is
+passed as a query string parameter or via an HTTP header.
+
+Setting up the authenticator is just 3 steps:
+
+* :ref:`cookbook-guard-api-authenticator`
+* :ref:`cookbook-guard-api-configuration`
+* :ref:`cookbook-guard-api-user-provider`
+
+.. _cookbook-guard-api-authenticator:
+
+A) Create the Guard Authenticator
+---------------------------------
+
+Suppose you want to read an ``X-API-TOKEN`` header on each request and use
+that to authenticate the user. To do this, create a class that extends
+:class:`Symfony\\Component\\Security\\Guard\\AbstractGuardAuthenticator`
+(or which implements :class:`Symfony\\Component\\Security\\Guard\\GuardAuthenticatorInterface`)::
+
+    // src/AppBundle/Security/TokenAuthenticator.php
+    namespace AppBundle\Security;
+
+    use Symfony\Component\HttpFoundation\Request;
+    use Symfony\Component\HttpFoundation\Response;
+    use Symfony\Component\Security\Core\Exception\AuthenticationCredentialsNotFoundException;
+    use Symfony\Component\Security\Guard\AbstractGuardAuthenticator;
+    use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+    use Symfony\Component\Security\Core\Exception\AuthenticationException;
+    use Symfony\Component\Security\Core\User\UserProviderInterface;
+
+    class TokenAuthenticator extends AbstractGuardAuthenticator
+    {
+        public function getCredentialsFromRequest(Request $request)
+        {
+            $token = $request->headers->get('X-AUTH-TOKEN');
+
+            // no token? Don't do anything :D
+            if (!$token) {
+                return;
+            }
+
+            return array(
+                'token' => $token,
+            );
+        }
+
+        public function authenticate($credentials, UserProviderInterface $userProvider)
+        {
+            $token = $credentials['token'];
+
+            // call a method on your UserProvider - see below for details
+            $user = $userProvider->loadUserByToken($token);
+
+            if (!$user) {
+                throw new AuthenticationCredentialsNotFoundException();
+            }
+
+            return $user;
+        }
+
+        public function onAuthenticationSuccess(Request $request, TokenInterface $token, $providerKey)
+        {
+            // on success, just let the request keep going!
+            return null;
+        }
+
+        public function onAuthenticationFailure(Request $request, AuthenticationException $exception)
+        {
+            $data = array(
+                // you might translate this message
+                'message' => $exception->getMessageKey()
```

Comment:
at least, data should be replaced: `str_replace($exception->getMessageData(), $exception->getMessageKey());`